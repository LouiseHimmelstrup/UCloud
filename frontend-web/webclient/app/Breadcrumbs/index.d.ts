export interface Breadcrumbs { currentPath: string, navigate: (path: string) => void, homeFolder: string }

export interface BreadCrumbMapping {
    actualPath: string
    local: string
}