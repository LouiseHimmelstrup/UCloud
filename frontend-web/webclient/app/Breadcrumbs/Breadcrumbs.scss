.ui.breadcrumb.breadcrumb-margin {
    margin: 10px 0px 30px 0px;
}