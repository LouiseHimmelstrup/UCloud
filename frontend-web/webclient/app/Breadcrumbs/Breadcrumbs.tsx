import * as React from "react";
import { Breadcrumb } from "semantic-ui-react";
import { BreadCrumbMapping, Breadcrumbs as BreadCrumbsList } from ".";
import { Cloud } from "Authentication/SDUCloudObject";

export const BreadCrumbs = ({ currentPath, navigate }: BreadCrumbsList) => {
    if (!currentPath) return null;
    const pathsMapping = buildBreadCrumbs(currentPath);
    const activePathsMapping = pathsMapping.pop()
    const breadcrumbs = pathsMapping.map((path, index) => (
        <React.Fragment key={index}>
            <Breadcrumb.Section onClick={() => navigate(`${path.actualPath}`)} link>
                {path.local}
            </Breadcrumb.Section>
            <Breadcrumb.Divider />
        </React.Fragment>
    ));

    return (
        <Breadcrumb size="large" className="breadcrumb-margin">
            {breadcrumbs}
            <Breadcrumb.Section active>
                {activePathsMapping ? activePathsMapping.local : null}
            </Breadcrumb.Section>
            <Breadcrumb.Divider />
        </Breadcrumb>
    );

}

export function buildBreadCrumbs(path: string) {
    const paths = path.split("/").filter((path: string) => path);
    let pathsMapping: BreadCrumbMapping[] = [];
    for (let i = 0; i < paths.length; i++) {
        let actualPath = "/";
        for (let j = 0; j <= i; j++) {
            actualPath += paths[j] + "/";
        }
        pathsMapping.push({ actualPath: actualPath, local: paths[i] });
    }
    if (path.includes(Cloud.homeFolder)) { // remove first two indices 
        pathsMapping =
            [{ actualPath: Cloud.homeFolder, local: "Home" }].concat(pathsMapping.slice(2, pathsMapping.length));
    }
    return pathsMapping;
}