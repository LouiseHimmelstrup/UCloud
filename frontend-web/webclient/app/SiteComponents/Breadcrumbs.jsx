import React from "react";
import {Breadcrumb} from "react-bootstrap";

export const BreadCrumbs = ({currentPath, navigate}) => {
    if (!currentPath) {
        return null;
    }
    let pathsMapping = buildBreadCrumbs(currentPath);
    let breadcrumbs = pathsMapping.map((path, index) =>
        <Breadcrumb.Item key={index} active={pathsMapping.length === index + 1} onClick={() => navigate(`${path.actualPath}`)}>{path.local}</Breadcrumb.Item>
    );
    return (
        <Breadcrumb>
            {breadcrumbs}
        </Breadcrumb>
    );

}

export function buildBreadCrumbs(path) {
    let paths = path.split("/").filter(path => path);
    let pathsMapping = [];
    for (let i = 0; i < paths.length; i++) {
        let actualPath = "/";
        for (let j = 0; j <= i; j++) {
            actualPath += paths[j] + "/";
        }
        pathsMapping.push({actualPath: actualPath, local: paths[i]})
    }
    return pathsMapping
}