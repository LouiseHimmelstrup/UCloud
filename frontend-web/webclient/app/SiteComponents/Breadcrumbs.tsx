import * as React from "react";
import "react-bootstrap";
import { BreadcrumbItem, Breadcrumb } from "react-bootstrap/lib";


interface Breadcrumbs { currentPath: string, navigate: (path: string) => void }
export const BreadCrumbs = ({ currentPath, navigate }: Breadcrumbs) => {
    if (!currentPath) {
        return null;
    }
    let pathsMapping = buildBreadCrumbs(currentPath);
    let breadcrumbs = pathsMapping.map((path, index) => (
        <BreadcrumbItem key={index} active={pathsMapping.length === index + 1}>
            <span onClick={() => navigate(`${path.actualPath}`)}>{path.local}</span>
        </BreadcrumbItem>
    ));
    return (
        <Breadcrumb>
            {breadcrumbs}
        </Breadcrumb>
    );

}

export function buildBreadCrumbs(path: string) {
    const paths: string[] = path.split("/").filter((path: string) => path);
    let pathsMapping = [];
    for (let i = 0; i < paths.length; i++) {
        let actualPath = "/";
        for (let j = 0; j <= i; j++) {
            actualPath += paths[j] + "/";
        }
        pathsMapping.push({ actualPath: actualPath, local: paths[i] })
    }
    return pathsMapping
}