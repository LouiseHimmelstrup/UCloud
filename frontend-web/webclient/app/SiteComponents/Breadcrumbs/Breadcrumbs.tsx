import * as React from "react";
import { Breadcrumb } from "semantic-ui-react";
import "./Breadcrumbs.scss";

interface Breadcrumbs { currentPath: string, navigate: (path: string) => void }
export const BreadCrumbs = ({ currentPath, navigate }: Breadcrumbs) => {
    if (!currentPath) {
        return null;
    }
    const pathsMapping = buildBreadCrumbs(currentPath);
    const activePathsMapping = pathsMapping.pop()
    const breadcrumbs = pathsMapping.map((path, index) => (
        <React.Fragment key={index}>
            <Breadcrumb.Section onClick={() => navigate(`${path.actualPath}`)} link>
                {path.local}
            </Breadcrumb.Section>
            <Breadcrumb.Divider />
        </React.Fragment>
    ));
    
    return (
        <Breadcrumb className="breadcrumb-margin">
            {breadcrumbs}
            <Breadcrumb.Section active>
                {activePathsMapping.local}
            </Breadcrumb.Section>
        </Breadcrumb>
    );

}

export function buildBreadCrumbs(path: string) {
    const paths = path.split("/").filter((path: string) => path);
    let pathsMapping = [];
    for (let i = 0; i < paths.length; i++) {
        let actualPath = "/";
        for (let j = 0; j <= i; j++) {
            actualPath += paths[j] + "/";
        }
        pathsMapping.push({ actualPath: actualPath, local: paths[i] });
    }
    return pathsMapping;
}