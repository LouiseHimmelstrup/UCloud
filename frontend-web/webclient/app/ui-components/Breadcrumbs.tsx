import * as React from "react";
import {injectStyle} from "@/Unstyled";

// https://www.w3schools.com/howto/howto_css_breadcrumbs.asp

export const BreadCrumbsClass = injectStyle("breadcrumbs", k => `
    ${k} {
        display: flex;
        width: calc(100% - 180px);
    }
    
    ${k}[data-embedded="true"] {
        width: calc(100% - 50px);
    }
    
    ${k} > span {
        font-size: 25px;
        display: inline-block;
        text-overflow: ellipsis;
        white-space: nowrap;
        overflow: hidden;
        
        color: var(--text, #f00);
        text-decoration: none;
    }
    
    ${k} > span:hover {
        cursor: pointer;
        color: var(--blue);
    }
    
    ${k} > span + span:before {
        padding: 0 8px;
        vertical-align: top;
        color: var(--text, #f00);
        content: "/";
    }
    
    ${k}.isMain > span:last-child:hover {
        color: var(--text, #f00);
        cursor: default;
    }
`);

export const BreadCrumbsBase: React.FunctionComponent<{
    embedded?: boolean;
    children?: React.ReactNode;
    className?: string;
}> = props => {
    return <div
        className={BreadCrumbsClass + " " + (props.className ?? "")}
        data-embedded={props.embedded === true}
        children={props.children}
    />;
}
