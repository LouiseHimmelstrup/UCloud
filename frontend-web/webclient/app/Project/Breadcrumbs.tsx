import * as React from "react";
import styled from "styled-components";
import Spinner, {HexSpinWrapper} from "@/LoadingIcon/LoadingIcon";
import {BreadCrumbsBase} from "@/ui-components/Breadcrumbs";
import {Link} from "@/ui-components";
import {useProject} from "./cache";
import {useProjectId} from "./Api";

export interface ProjectBreadcrumbsProps {
    crumbs: {title: string, link?: string}[];
    allowPersonalProject?: boolean;
    omitActiveProject?: boolean;
}

const ProjectBreadcrumbsWrapper = styled(BreadCrumbsBase)`
    width: 100%;
    max-width: unset;
    flex-grow: 1;
    margin-bottom: 12px;
    
    ${HexSpinWrapper} {
        margin: 0;
        display: inline;
    }
`;

export const ProjectBreadcrumbs: React.FunctionComponent<ProjectBreadcrumbsProps> = props => {
    const projectId = useProjectId();
    const project = useProject();
    let projectNameComponent = <Spinner />;
    if (project.fetch().id !== "") {
        const title = project.fetch().specification.title;
        projectNameComponent = <>{title.slice(0, 20).trim()}{title.length > 20 ? "..." : ""}</>;
    }

    return <ProjectBreadcrumbsWrapper embedded={false}>
        <span>My Projects</span>
        {projectId && !props.omitActiveProject ? (
            <span>{projectNameComponent}</span>
        ) : props.allowPersonalProject ? <span>My Workspace</span> : null}
        {props.crumbs.map((crumb, idx) => {
            if (crumb.link) {
                return <span key={idx}><Link to={crumb.link}>{crumb.title}</Link></span>;
            } else {
                return <span key={idx}>{crumb.title}</span>;
            }
        })}
    </ProjectBreadcrumbsWrapper>;
};
