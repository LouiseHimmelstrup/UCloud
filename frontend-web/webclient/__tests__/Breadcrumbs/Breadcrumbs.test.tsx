import * as React from "react";
import * as Renderer from "react-test-renderer";
import { BreadCrumbs } from "Breadcrumbs/Breadcrumbs";
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { Breadcrumb } from "semantic-ui-react";

configure({ adapter: new Adapter() });

describe("Breadcrumbs", () => {
    const homefolder = "/home/tester@user.dk/";
    test("Build breadcrumbs from empty string", () => {
        expect(Renderer.create(
            <BreadCrumbs
                currentPath=""
                navigate={() => null}
                homeFolder={homefolder}
            />).toJSON()).toMatchSnapshot()
    });

    test("Build breadcrumbs from string without home in it", () => {
        expect(Renderer.create(
            <BreadCrumbs
                currentPath="Not/Home/Folder"
                navigate={() => null}
                homeFolder={homefolder}
            />).toJSON()).toMatchSnapshot()
    });

    test("Build breadcrumbs from string with home in it", () => {
        expect(Renderer.create(
            <BreadCrumbs
                currentPath={`${homefolder}followed/by/something/more`}
                navigate={() => null}
                homeFolder={homefolder}
            />).toJSON()).toMatchSnapshot()
    });

    test("Clicking navigate on active breadcrumb", () => {
        const navigate = jest.fn();
        const node = shallow(<BreadCrumbs currentPath={"/home/else/where/"} navigate={navigate} homeFolder={homefolder}/>);
        expect(navigate).toBeCalledTimes(0);
        node.findWhere(it => it.type() === Breadcrumb.Section).first().simulate("click")
        expect(navigate).toHaveBeenCalled();
    });
});