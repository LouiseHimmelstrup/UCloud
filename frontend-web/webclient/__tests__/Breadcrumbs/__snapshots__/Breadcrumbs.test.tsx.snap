// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs Build breadcrumbs from empty string 1`] = `null`;

exports[`Breadcrumbs Build breadcrumbs from string with home in it 1`] = `
<div
  className="ui large breadcrumb breadcrumb-margin"
>
  <a
    className="section"
    onClick={[Function]}
  >
    Home
  </a>
  <div
    className="divider"
  >
    /
  </div>
  <a
    className="section"
    onClick={[Function]}
  >
    followed
  </a>
  <div
    className="divider"
  >
    /
  </div>
  <a
    className="section"
    onClick={[Function]}
  >
    by
  </a>
  <div
    className="divider"
  >
    /
  </div>
  <a
    className="section"
    onClick={[Function]}
  >
    something
  </a>
  <div
    className="divider"
  >
    /
  </div>
  <div
    className="active section"
    onClick={[Function]}
  >
    more
  </div>
  <div
    className="divider"
  >
    /
  </div>
</div>
`;

exports[`Breadcrumbs Build breadcrumbs from string without home in it 1`] = `
<div
  className="ui large breadcrumb breadcrumb-margin"
>
  <a
    className="section"
    onClick={[Function]}
  >
    Not
  </a>
  <div
    className="divider"
  >
    /
  </div>
  <a
    className="section"
    onClick={[Function]}
  >
    Home
  </a>
  <div
    className="divider"
  >
    /
  </div>
  <div
    className="active section"
    onClick={[Function]}
  >
    Folder
  </div>
  <div
    className="divider"
  >
    /
  </div>
</div>
`;
